language-design.md

```diff
@@ -59,50 +72,45 @@ What does the post on grayscale tell us about the process of API design?
-> language, like any other programming language. … even small changes to the
-> script may introduce subtle syntactic errors that baffle users. It is easy to
-> read AppleScript, but quite hard to write it.
+The Pavlus article mentions the researchers' comments that people preferred "natural-language replacements for some of the more abstruse syntax". In other words, people found it easier to work with code that looks more like a human language (e.g., English). Consider the following quote by William R. Cook, one of the creators of AppleScript:
+
+> The experiment in designing a language that resembled natural languages (English and Japanese) was not successful. It was assumed that scripts should be presented in “natural language” so that average people could read and write them. ... In the end the syntactic variations and flexibility did more to confuse programmers than to help them out. It is not clear whether it is easier for novice users to work with a scripting language that resembles natural language, with all its special cases and idiosyncrasies. The main problem is that AppleScript only appears to be a natural language: in fact, it is an artificial language, like any other programming language. … even small changes to the script may introduce subtle syntactic errors that baffle users. It is easy to read AppleScript, but quite hard to write it.
 [[Cook 2007, page 1-20]](https://dl.acm.org/citation.cfm?doid=1238844.1238845)
 
-Are these two experiences of natural languages at odds with one another? Would
-you choose to include natural language in the design of a DSL? If so, how might
-you do so? If not, why not?
+Are these two experiences of natural languages at odds with one another? Would you choose to include natural language in the design of a DSL? If so, how might you do so? If not, why not?
 
 **Response**
 
+There are two ways to look at this dichotomy between programming languages which have the feel of a "natural language" being both better and worse for new programmers.
 
+We may note that there is a spectrum to how closely the language resembles natural language. On one end, we have the binary / hexidecimal codes in which computers store their instructions, and on the other side we have languages, like AppleScript, which *read* much like a natural language. One might note that both of these extremes are poor choices for programmers. The former, while entirely precise and unambiguous is not structured, nor is it easy to see patterns or errors.  The latter on the other hand depends on the natural languages, which are full of their own ambiguities and exceptional grammars. Thus, one might argue that the correct programming language is somewhere in between these two extrema, and the language tested in the Palvus article may have been at a good location along this spectrum, while a language like AppleScript was too close to the ambiguous, hard-to-tell-what's-a-keyword, syntax of natural languages.
```

Comment:
Thinking of the languages on a spectrum is a very good point, especially that there are tradeoffs everywhere along the spectrum to consider. Also, the point of "exception grammars" that characterize natural languages is very important. Anthony and I discussed how it is helpful to have languages that have some natural language essence but follow a very strict, unique grammar (which may come with the tradeoff with not sounding entirely natural). The "hard-to-tell-what's-a-keyword" is the most annoying part of some languages&mdash;especially in AppleScript. Although patterns in this language exist, they can be broken depending on the writer of the code&mdash;something I'd argue Steele would not think is a positive of these languages.